src/pages/experiments/activities/index.js

```diff
@@ -1,32 +1,85 @@
 import React from 'react';
+import { useState } from 'react';
+import { useEffect } from 'react';
+import { useCallback } from 'react';
 import { connect } from 'react-redux';
 
 import FetchAlert from '../../../components/fetch-alert';
 import Heatmap from './heatmap';
 import Controls from './controls';
+import { actionStatuses } from '../../../actions/fetch';
 import { isString } from '../../../util/types';
 import { isArray } from '../../../util/types';
+import { uniqueMap } from '../../../util/object';
+
+/* eslint import/no-webpack-loader-syntax: off */
+import worker from 'workerize-loader!./cluster';
 
 import './index.css';
 
 // sample activities section
 
-let Activities = ({ activities }) => (
-  <>
-    {isString(activities) && (
-      <FetchAlert status={activities} subject='activities' />
-    )}
-    {isArray(activities) && (
-      <>
-        <Heatmap />
-        <Controls />
-      </>
-    )}
-  </>
-);
+const workerInstance = worker();
+
+let Activities = ({ activities }) => {
+  const [clusteredSamples, setClusteredSamples] = useState(null);
+  const [clusteredSignatures, setClusteredSignatures] = useState(null);
+
+  let samples;
+  let signatures;
+  if (isArray(activities)) {
+    samples = uniqueMap(activities, (activity) => activity.sample);
+    signatures = uniqueMap(activities, (activity) => activity.signature);
+  }
+
+  useEffect(() => {
+    setClusteredSamples(null);
+    setClusteredSignatures(null);
+  }, [activities]);
+
+  const clusterSamples = useCallback(async () => {
+    setClusteredSamples(actionStatuses.LOADING);
+    setClusteredSamples(
+      await workerInstance.clusterData(activities, 'sample', 'value')
+    );
+  }, [activities]);
+
+  const clusterSignatures = useCallback(async () => {
+    setClusteredSignatures(actionStatuses.LOADING);
+    setClusteredSignatures(
+      await workerInstance.clusterData(activities, 'signature', 'value')
+    );
+  }, [activities]);
+
+  return (
+    <>
+      {isString(activities) && (
+        <FetchAlert status={activities} subject='activities' />
+      )}
+      {isArray(activities) && (
+        <>
+          <Heatmap
+            activities={activities}
+            samples={isArray(clusteredSamples) ? clusteredSamples : samples}
+            signatures={
+              isArray(clusteredSignatures) ? clusteredSignatures : signatures
+            }
+          />
+          <Controls
+            activities={activities}
+            clusteredSamples={clusteredSamples}
```

Comment:
Do you mind making the variable names `clusteredSamples` and `clusterSamples` more distinct? At the first sight I thought they were identical. Same thing with `clusteredSignatures`.

Reply:
What do you suggest. I want to be very accurate because there are samples, unclusteredSamples, and clusteredSamples. I don't want to be inaccurate or imprecise. It is unfortunate that the english language can be so ambiguous.

Reply:
Is `clusterSamples` and `clusterSignatures` functions? If they are, what about renaming them to `makeSamplesCluster` and `makeSignaturesCluster`?

Reply:
They are functions yes. But those descriptions aren't accurate. The functions basically sort the samples/signatures based on clustering.

Reply:
I can use "sorted" instead of "clustered" but then I would want to be consistent and do `sortedSamples` and `sortSamples` anyway.

Reply:
I do think "sort..." is better.

Reply:
Okay, I've changed it to sort. Technically sort is slightly more accurate because the algorithm is clustering the data, then returning a list of sample/signature ids sorted based on those clusters.

Reply:
:+1: